### PR TITLE
Computation of scaling parameters for input features

### PR DESCRIPTION
**Pull request by yaourtpourtoi**
This is a PR which introduces the module to compute scaling parameters (currently `mean`, `std`, `lim_min` and `lim_max`) for input features to DeepTau. There are 5 additions to the current state of the repo:

* `Training/python/feature_scaling.py`: main script which handles the computation
* `Training/python/scaling_utils.py`: contains helper functions used by `feature_scaling.py`
*  `Training/notebooks/2017v2/scaling_validation.ipynb`: Jupyter notebook which produces validation plots as presented in [these slides](https://indico.cern.ch/event/1038235/contributions/4359970/attachments/2242374/3803434/scaling_update_11May.pdf). Can be used for spotting anomalous values and validating input files' uniformity.
*  `Training/configs/training_v1.yaml`: augmented version of the current cfg file. Introduced updates don't change the file's structure but rather add more information, which is further commented in the corresponding parts of the file. 
*  `Training/configs/scaling_params_v0.json`: test json file with scaling parameters produced by processing `/eos/cms/store/group/phys_tau/TauML/prod_2018_v2/ShuffleMergeSpectral_v1/ShuffleMergeSpectral_0.root` (1 file)

In order to run the calculation on all variable types use:
```python 
python feature_scaling.py --cfg ../configs/training_v1.yaml --var_types -1
```
This will compute the the scaling parameters for all variable type and features wherein as listed in `training_v1.yaml`. Here the flags specify
* `--cfg`: path to `training_v1.yaml` file
* `--var_types`: set of variable types from `training_v1.yaml` (e.g. `Electron PfCand_nHad`) to be run on. Default = -1 (all which are listed in the cfg file)

After computation is finished, the output json files with scaling parameters will be stored as described in `Scaling_setup` section of `training_v1.yaml` file (`output_json_folder` field). There, one can also specify a set of files to be run on and the way logging is handled.  

There are a few remaining items to be done:

- [x] integrate changes to `training_v1.yaml` with DataLoader (i.e. update the parsing method)
- [x] integrate changes to output json with DataLoader (i.e. update the parsing method)
- [x] perform NaN/inf/large or small values cleanup in `feature_scaling.py` (current `scaling_params_v0.json` has 2 NaNs) 
- [x] propagate the same processing to DataLoader
- [x] ~implement job submission using `law` to be able to run on all the files within reasonable amount of time~
- [x] check that the values in the final json file are complete and reasonable (e.g. compare with [previous values](https://github.com/cms-tau-pog/TauMLTools/blob/0a8fad233727b4fcf4d41afdabe90b2922b69902/Analysis/bin/TrainingTupleProducer.cxx) from `TrainingTupleProducer.cxx`)

### CODE REVIEW

**Comment by yaourtpourtoi** [comment]
UPD: implemented inf masking (removes previously obtained NaNs) + integrated updated json (global field+readability) and yaml cfg formats with `DataLoader_test.py` parsing methods + minor fixes

File: Training/configs/training_v1.yaml
Comment (kandrosov) [inline Training/configs/training_v1.yaml:217]: don't scale `pfCand_*_pvAssociationQuality` - it is categorical in nature and we can consider using an embedding layer for it (and other categorical features that could be added in the future).

File: Training/notebooks/2017v2/scaling_validation.ipynb
Comment (kandrosov) [inline Training/notebooks/2017v2/scaling_validation.ipynb:1]: I suggest adding the following validation test: fraction of entries outside of the validity range, which is specified in yaml (e.g. [-5, 5]) should be below a given threshold. Variables for which test fails should be reported and we need to consider a case-by-case basis. As a threshold, I propose to use 0.01.

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Training/notebooks/2017v2/scaling_validation.ipynb
Comment (yaourtpourtoi) [reply Training/notebooks/2017v2/scaling_validation.ipynb:1]: @kandrosov Sounds good, maybe I could propose the following way, which is slightly different from what you suggest but would have a benefit of not rerunning the computation once again. Since in the scaling code I loop firstly over files and then over features, what I could do is for each feature array during aggregation store (per file) the values of it's 0.01 and 0.99 percentiles. This would already give us some intuition on the range of the distribution and also the dynamics of its variation from one file to the other. And this also should be rather straight-forward to implement. Although there ofc can be low statistics for some of the features and percentile estimates will be stochastic... Would it make sense to try it out? Otherwise surely I can implement it the way you suggested.

**Comment by yaourtpourtoi** [comment]
Updating PR with the following items:

- quantile validation as was shown on [this meeting](https://indico.cern.ch/event/1044740/contributions/4389426/attachments/2255446/3827568/scaling_update_1June.pdf):
    -  corresponding plots can be produced with 
                `python quantile_validation.py --train_cfg ../configs/training_v1.yaml --scaling_file ../configs/scaling_params_v1.json --quantile_file ../configs/quantile_params_v1_fid_0.json --output_folder quantile_plots --file_id 0`
    - `notebooks/2017v2/scaling_validation.ipynb` was updated with a Jupyter-like version of this script
    - since the script produces by default only "suspicious" plots, [here](https://cernbox.cern.ch/index.php/s/L0GCIin6KSQczSE) or in  `/afs/cern.ch/work/o/ofilatov/public/scaling_v1/all_var_types/quantile_plots/` I made quantile plots for all the linear/normal features for all 5 files, so that dependance of quantiles across files can also be compared (briefly studying seems to be reasonably similar) along with the ranges themselves
- finalised synchronisation of updates to `training_v1.yaml` and `scaling_params_v1.json` with `DataLoader_test.py` updated from @shedprog's fork: 
    - to avoid warnings line `TAU_IDS()` needs to be removed and line `#include "TauMLTools/Analysis/interface/TauIdResults.h"` commented in `Analysis/interface/TauTuple.h`(did not commit this change)
    - testing `python DataLoader_test.py` doesn't produce any errors and printout shows batches are being loaded. 
- updated json files with scaling (`configs/scaling_params_v1.json`) and quantile parameters (`configs/quantile_params_v1_fid_0.json`):
    -  these are produced by running `python feature_scaling.py --cfg ../configs/training_v1.yaml --var_types -1` with `file_range` set to [0, 5] (first 5 S&M files in the sorted list)
    - `quantile_params_v1_fid_0.json` corresponds to quantiles calculated on the first file in the specified [0, 5] file range (file id = 0)
    - more detailed logging can found on [cernbox](https://cernbox.cern.ch/index.php/s/zIEqm0TM2co8Jty) or in `/afs/cern.ch/work/o/ofilatov/public/scaling_v1/` 
- updates following the comments from discussion above:
    - introduced categorical flag into `training_v1.yaml`: behaviour for such labelled variables is the same as for "no_scaling" ones
    - removed `abs()` in `pfCand_particleType ` 
- minor fixes and improvements

**Comment by yaourtpourtoi** [comment]
But since my fork fell behind from the main repo, I had to cherry-pick `DataLoader_main.h` in 010db10 so that `DataLoader_test.py` works out of the box on my master - which might be worthwhile reverting

File: Training/configs/training_v1.yaml
Comment (yaourtpourtoi) [reply Training/configs/training_v1.yaml]: fixed in 8c4362d

File: Training/configs/training_v1.yaml
Comment (yaourtpourtoi) [reply Training/configs/training_v1.yaml]: added categorical flag in a062a86

**Comment by shedprog** [comment]
Is it possible to convert ipynb->py or at least to remove meta data, because it seems that amount of code is unreviewable?

**Review by kandrosov — CHANGES_REQUESTED** [review]
Please, resolve conflicts.

File: Training/configs/training_v1.yaml
Comment (yaourtpourtoi) [reply Training/configs/training_v1.yaml:169]: Hm, I am thinking about this and from what I understand there might be essentially no difference between normal and linear scaling in those exceptions. Basically, they are both implemented as a linear transformation followed by clamping, and the only difference is how the range of clamping is specified in the training cfg. In linear case it is assumed to be "a priori known" (and hence explicitly written, like [1., 5.2]), while for normal scaling it is sort of not so we leave it to a suppossedly-large-enough-to-cover "+-5 std range". But then both of them might turn out to not entirely cover the whole distribution. For example, with linear case as it is in `tau_E_over_pt`, and for the normal case still 5 std can be not enough, because distributions are mostly not gaussian-like. So in the end there are still checks to be made with actual quantiles if we want to be sure. And if they reveal undercoverage for clamping, then for linear case it can be either adjusted by extending the range in cfg, or computing the mean/std and still adjusting the range but now in terms of std units, so it is basically the same thing :) That is why I am a bit confused, because we can simply respecify in the training cfg [1., 5.2] -> [1., 6.2] and that would do the job if excessive clamping is the concerning part.

Comment (yaourtpourtoi) [reply Training/configs/training_v1.yaml:169]: for the reference, this is how `tau_E_over_pt` distribution will look like after applying scaling with current parameters. Top: original distribution, bottom: transformed
image

Comment (yaourtpourtoi) [reply Training/configs/training_v1.yaml:169]: and this is the corresponding quantile plot
image

Comment (kandrosov) [reply Training/configs/training_v1.yaml:169]: Thank you for the details! I think, as a general practice, we should try to avoid having "custom" intervals in the config because it is less robust for possible changes. In `tau_E_over_p` case I think it is safe to switch to the normal transformation so that we don't need to manually control for each S&M production that the range is still relevant.

In the future, we can think about more automated range determination, because as you mentioned for some very non-gaussian distributions the (-5 * std, +5 * std) interval doesn't cover 95% of events, so clamping at 5 could be a bit suboptimal. But for the first round, I think we are fine with what we have in place + few minor tunings listed in the PR comments.

File: Training/interface/DataLoader_main.h
Comment (yaourtpourtoi) [reply Training/interface/DataLoader_main.h]: done in 75c4699 (general fix of feature disabling) and 5de595e

File: Training/configs/training_v1.yaml
Comment (yaourtpourtoi) [reply Training/configs/training_v1.yaml]: done in 4bf0314

File: Training/configs/training_v1.yaml
Comment (yaourtpourtoi) [reply Training/configs/training_v1.yaml]: updated in 2a1ec38

**Comment by yaourtpourtoi** [comment]
Updates resolve comments from the conversation above plus include additional items discussed during the meeting:
* copied `/afs/cern.ch/work/m/myshched/public/ShuffleSpectrum.root` -> `/eos/cms/store/group/phys_tau/TauML/prod_2018_v2/ShuffleSpectrum.root`
* `tau_E_over_pt` changed to "normal" type of scaling + update its mean/std in json 
* fix of feature disabling + disabled `tau_phi`: previously the code wasn't working properly
* as I was running the scaling tool, it spotted few more infs in variables, which I added a check for in `DataLoader_main.h`
    * `pfCand_chHad_vertex_dz`
    * `pfCand_chHad_vertex_dz_tauFL`
    * `pfCand_chHad_dz`
    * `pfCand_chHad_dz_sig`
* following @shedprog suggestion splitted the scaling validation notebook into two scripts `python/plot_scaling_convergence.py` and `python/plot_quantile_ranges.py`
* minor fixes and improvements

After merging with upstream master (mine was out of synch from Mykyta's and Luca's updates), running `python DataLoader_test.py` works out-of-the-box: printouts that batches are being formed except for warnings "*: x and y axes are switched!" and "Empty tuple detected! element= 12".

**Review by kandrosov — APPROVED** [review]

File: Training/configs/training_v1.yaml
Comment (shedprog) [inline Training/configs/training_v1.yaml:222]: @yaourtpourtoi @kandrosov if I did not mix it up, I have a question about the following variables: `pfCand_pvAssociationQuality`, `pfCand_lostInnerHits` and also to the `pfCand_fromPV` which we use for the muon collection: did we agree that all variables should be scaled to [-1, +1]? In this I case, we need to use linear mapping on this interval (because these variables are not distributed within [-1, +1]), yes it is categorical, but we are not handling it appropriately on the NN input (not embedding them or something), so wouldn't this feature just negatively influence the scale inside NN?

For instance looking at the `pfCand_pvAssociationQuality`:
image

Comment (yaourtpourtoi) [reply Training/configs/training_v1.yaml:222]: Yes, these variables as of now definitely go beyond the [-1, 1] scale. However, it's hard for me personally to predict intuitively whether this can cause troubles for the training/overall performance. For example, `pfCand_pvAssociationQuality` seems to have mostly values 8, so I can speculate that this feature might be considered as trivially redundant by the model (and therefore the different scale doesn't cause problems). So I'd personally ran experiments to see if the scaling of those features changes something in terms of performance/convergence (or maybe even if the scaling is needed at all..). And additionally, it would be interesting to visualise the input tensor (like a 3D heatmap) as it goes through the first layers of the model to see, how the inputs with "out-of-scale" values propagate through the network - e.g. if the amplitude is getting magnified or on the contrary diluted back to average scale.

**Pull request merged**
